### DISCUSSION TITLE
Delayed output variable defined in $MAIN

### DISCUSSION THREAD

**Discussion by walsu** [metrumresearchgroup/mrgsolve#1044] [Q&A] [answered]
Hello,
I just found that the output variable defined in $MAIN is delayed by one timepoint unlike the value defined in other places ($GLOBAL, $ODE, $TABLE). Is this already an expected result? I searched for the manual, but couldn't find it, so I'm writing here with a reproducible example.

```
library(tidyverse)
library(mrgsolve)

code <- '
$PARAM @annotated
TVCL   : 0.2   : Clearance (L/day)
TVVC   : 15    : Central Volume (L)
TVKA   : 0.5   : Absorption rate constant (1/day)

$CMT @annotated
DEPOT  : Depot compartment 
CENT   : Central compartment (mass)

$MAIN
double CL = TVCL * exp(ETA(1));
double VC = TVVC * exp(ETA(2));
double KA = TVKA * exp(ETA(3));

double CP_MAIN = CP*10 ;

$GLOBAL
#define CP (CENT/VC)
#define DV (CP*(1+EPS(1)))
#define CP_GLOBAL (CP*10)

$OMEGA 0.1 0.1 0.1 
$SIGMA 0.01

$ODE
dxdt_DEPOT = -KA*DEPOT;
dxdt_CENT =   KA*DEPOT -(CL/VC)*CENT;
double CP_ODE = CP*10 ;

$TABLE 
capture CP_TABLE = CP*10;

$CAPTURE @annotated
CP : IPRED (ng/mL)
DV : Observed (ng/mL)
CP_MAIN : CP defined in $MAIN 
CP_GLOBAL : CP defined in $GLOBAL
CP_ODE : CP defined in $ODE
'

mod <- mcode("demo", code)

df <- mod %>% 
  zero_re() %>%
  ev(amt=1) %>% 
  carry_out(DOSE, evid, amt, CP_MAIN, CP_GLOBAL, CP_ODE) %>% 
  mrgsim(delta=1, end=5) %>% 
  as_tibble()

df %>% 
  select(time, CP, CP_MAIN, CP_GLOBAL, CP_ODE, CP_TABLE)
```
![image](https://user-images.githubusercontent.com/9631440/214345705-514e6a92-b220-450b-a588-4d27b1952ed5.png)

**Comment by kylebaron** [comment]
Hi @walsu  - 
Yes, this is expected. When you calculate `CP` in `$MAIN`, you are using the amount in the central compartment prior to advancing the system. When you calculate it in `$ODE` or `$TABLE`, you'll get the amount in the central compartment after the system advances (we make one extra call to `$ODE` after system advance to make that happen for `$ODE`). In the example, `CP` isn't really calculated in `$GLOBAL`; there, you are just defining a preprocessor directive for `CP_GLOBAL` and when you capture that, it's actually getting calculated in `$TABLE`. 

mrgsolve will let you access compartment amounts in `$MAIN` (or `$PK`), but I think for most applications, you want to calculate that in `$TABLE`. 

Some discussion of this here:
https://mrgsolve.org/user_guide/model-components.html#functions

Kyle

**Comment by walsu** [comment, marked as answer]
Thank you @kylebaron 
I remember reading that part of the manual before, but I didn't think it was very important at that time. 
Now it's very helpful to understand what is happening in the simulation step. 

Wansu

**Comment by kylebaron** [comment]
No problem. I think mentioning this implication specifically (accessing a compartment in $MAIN) could help clarify and I'll add that to the user guide.
